This is a collection of patches that make the GRUB2 build work on mingw32.  
They've been submitted upstream, but not yet accepted.

Index: grub2-1.99/grub-core/kern/emu/misc.c
===================================================================
--- grub2-1.99.orig/grub-core/kern/emu/misc.c	2011-09-06 22:25:48.157785280 -0500
+++ grub2-1.99/grub-core/kern/emu/misc.c	2011-09-06 22:25:53.149810026 -0500
@@ -232,6 +232,11 @@
 canonicalize_file_name (const char *path)
 {
   char *ret;
+#ifdef __MINGW32__
+  ret = xmalloc(PATH_MAX);
+  if (!_fullpath(ret, path, PATH_MAX))
+    return NULL;
+#else
 #ifdef PATH_MAX
   ret = xmalloc (PATH_MAX);
   if (!realpath (path, ret))
@@ -239,6 +244,7 @@
 #else
   ret = realpath (path, NULL);
 #endif
+#endif /* __MINGW32__ */
   return ret;
 }
 
Index: grub2-1.99/include/grub/util/misc.h
===================================================================
--- grub2-1.99.orig/include/grub/util/misc.h	2011-09-06 22:25:27.737684020 -0500
+++ grub2-1.99/include/grub/util/misc.h	2011-09-06 22:25:33.993715034 -0500
@@ -47,7 +47,6 @@
 
 void sync (void);
 int fsync (int fno);
-void sleep(int s);
 
 grub_int64_t grub_util_get_disk_size (char *name);
 
Index: grub2-1.99/util/misc.c
===================================================================
--- grub2-1.99.orig/util/misc.c	2011-09-06 22:25:22.085655995 -0500
+++ grub2-1.99/util/misc.c	2011-09-06 22:26:21.941952796 -0500
@@ -55,6 +55,7 @@
 #ifdef __MINGW32__
 #include <windows.h>
 #include <winioctl.h>
+#include "dirname.h"
 #endif
 
 #ifdef GRUB_UTIL
@@ -316,17 +317,13 @@
   return 0;
 }
 
-void sleep (int s)
-{
-  Sleep (s * 1000);
-}
-
 grub_int64_t
 grub_util_get_disk_size (char *name)
 {
   HANDLE hd;
   grub_int64_t size = -1LL;
 
+  strip_trailing_slashes(name);
   hd = CreateFile (name, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE,
                    0, OPEN_EXISTING, 0, 0);
 
